[WIP] Bars and downbeats for common time signature.

hacksdump opened the pull request:
Display and edit downbeats on the scrolling waveform.

Task List:

- [x] Display downbeats on the waveform.

- [x] Edit downbeat with a context menu.

This PR will be covering a part of the first phase tasks primarily as described in the planning page: [https://www.mixxx.org/wiki/doku.php/measures_downbeats_bars_and_phrases](https://www.mixxx.org/wiki/doku.php/measures_downbeats_bars_and_phrases)

hacksdump committed:
waveformbeat: extract beat object

hacksdump committed:
beatiterator: return beats proto object

hacksdump committed:
waveformbeat: draw for vertical orientation

hacksdump committed:
beat: add beat subclass of beats proto

hacksdump committed:
beat: add utility function to convert to sample position

hacksdump committed:
wbeatmenu: change downbeat via context menu

hacksdump committed:
wbeatmenu: fix cursor

crisclacerda commented:
Just skimmed through it and it looks like your redoing a lot of the work that was done on #2512, shouldn't you be looking into integrating your work into that?
@JaviVilarroig has done enormous work to get rid of the references to the old beat grid and I think we should use that. 
I also have done work to make beatmap behaves well for constant tempo and there is no point in having both beatgrid and beatmaps, so there is no more point for using beatgrid.
If anything focus your works only on beatmaps on master and then integrates it into the new beats class which is mostly the old beatmap. That's what I am doing.

hacksdump committed:
beatmap: fix broken test due to shared pointer

hacksdump committed:
dlgprefwaveform: add beatgrid mode choices

hacksdump committed:
dlgprefwaveform: control downbeat display with preferences

Be-ing commented on CMakeLists.txt:
```diff
@@ -544,6 +544,7 @@ add_library(mixxx-lib STATIC EXCLUDE_FROM_ALL
   src/sources/soundsourceproxy.cpp
   src/sources/soundsourcesndfile.cpp
   src/track/albuminfo.cpp
+  src/track/beat.cpp
```
Until #2777 is merged, also add new files to build/depends.py

uklotzde commented on src/track/beat.h:
```diff
@@ -0,0 +1,46 @@
+
+class Beat : public mixxx::track::io::Beat {
+  public:
+    Beat(){};
```
= default

hacksdump committed:
wbeatmenu: dynamic options based on preferences

uklotzde commented on src/track/beatgrid.cpp:
```diff
@@ -248,8 +262,8 @@ bool BeatGrid::findPrevNextBeats(double dSamples,
 
-
-std::unique_ptr<BeatIterator> BeatGrid::findBeats(double startSample, double stopSample) const {
+std::unique_ptr<BeatIterator> BeatGrid::findBeats(
```
Future (after #2512 has been merged): If BeatGrid implements a move constructor there is no need to return a new instance by pointer.

If possible we should use values instead of shared/unique pointers, especially for small, lightweight objects.

uklotzde commented on src/track/beatgrid.cpp:
```diff
@@ -365,4 +385,8 @@ void BeatGrid::setBpm(double dBpm) {
     emit updated();
 }
 
+void BeatGrid::setDownbeatOffset(int offset) {
```
In which unit is `offset` measured? Please always add Frames/Samples/Millis/Beats as a suffix.

uklotzde commented on src/track/beatmap.cpp:
```diff
@@ -26,8 +25,8 @@ inline double framesToSamples(const int frames) {
 
-bool BeatLessThan(const Beat& beat1, const Beat& beat2) {
-    return beat1.frame_position() < beat2.frame_position();
+bool BeatLessThan(const BeatPointer& beat1, const BeatPointer& beat2) {
```
As already mentioned we should avoid shared pointers if possible

uklotzde commented on src/track/beatmap.cpp:
```diff
@@ -38,19 +37,19 @@ class BeatMapIterator : public BeatIterator {
     }
 
-    virtual bool hasNext() const {
+    virtual bool hasNext() const override {
```
`virtual` is redundant if `override` is specified

uklotzde commented on src/track/beatmap.cpp:
```diff
@@ -38,19 +37,19 @@ class BeatMapIterator : public BeatIterator {
 
-    virtual double next() {
-        double beat = framesToSamples(m_currentBeat->frame_position());
+    virtual BeatPointer next() override {
```
Return by value (move)

uklotzde commented on src/track/beatmap.cpp:
```diff
@@ -136,18 +135,25 @@ void BeatMap::createFromBeatVector(const QVector<double>& beats) {
-            beat.set_frame_position(beatpos);
+            beat->set_frame_position(beatpos);
+            beat->setIndex(beatCount);
+            if (beatCount % 4 == 0) {
```
Define a `constexpr kDefaultBeatsPerMeasure` instead of using 4 literally

uklotzde commented on src/track/beatmap.cpp:
```diff
@@ -566,46 +572,46 @@ void BeatMap::scale(enum BPMScale scale) {
-        Beat beat;
+        int distance = it->get()->frame_position() - prevBeat->frame_position();
+        BeatPointer beat = BeatPointer(new Beat());
         for (int i = 1; i <= 3; i++) {
```
Both literals 3 and 4 need to be replaced in terms of kDefaultBeatsPerMeasure. Please check all the code for these literals and implicit assumptions.

hacksdump replied:
This is quadruple scaling, which means we insert three beats between two adjacent beats. It does not depend on Time Signature. However, downbeat adjustment will make use of `kDefaultBeatsPerMeasure` while scaling.

uklotzde commented on src/track/beatmap.h:
```diff
@@ -10,13 +10,14 @@
 #define BEAT_MAP_VERSION "BeatMap-1.0"
 
-typedef QList<mixxx::track::io::Beat> BeatList;
+typedef QList<BeatPointer> BeatList;
```
No pointers, just a list of values.

uklotzde commented on src/track/beatmap.h:
```diff
@@ -85,14 +86,16 @@ class BeatMap final : public Beats {
 
-    double calculateBpm(const mixxx::track::io::Beat& startBeat,
-                        const mixxx::track::io::Beat& stopBeat) const;
+    double calculateBpm(const BeatPointer& startBeat,
```
Values, not pointers

uklotzde commented on src/track/beats.h:
```diff
@@ -23,7 +24,7 @@ class BeatIterator {
     virtual ~BeatIterator() {}
     virtual bool hasNext() const = 0;
-    virtual double next() = 0;
+    virtual BeatPointer next() = 0;
```
Pass/move by value, no pointer

uklotzde commented on src/waveform/renderers/waveformbeat.cpp:
```diff
@@ -0,0 +1,64 @@
+#include <math.h>
```
Use "util/math.h"

uklotzde commented on src/waveform/renderers/waveformbeat.cpp:
```diff
@@ -0,0 +1,64 @@
+#include <math.h>
+#include <waveform/renderers/waveformbeat.h>
```
This is not a system include, i.e. "waveform/renderers/waveformbeat.h"

uklotzde commented on src/waveform/renderers/waveformbeat.cpp:
```diff
@@ -0,0 +1,64 @@
+}
+
+void WaveformBeat::draw(QPainter* painter) const {
+    const float edgeLength = 10;
```
Define a constexpr in an anonymous namespace at the top of the file

uklotzde commented on src/waveform/renderers/waveformbeat.cpp:
```diff
@@ -0,0 +1,64 @@
+    QPointF thirdPoint = baseMidPoint;
+
+    float halfBase = edgeLength / 2;
+    float altitude = sqrt(3) / 2 * edgeLength;
```
Move calculation into an inline function in the anonymous namespace

uklotzde commented on src/waveform/renderers/waveformbeat.h:
```diff
@@ -0,0 +1,42 @@
+#pragma once
+
+#include <proto/beats.pb.h>
```
"proto/beats.pb.h"

uklotzde commented on src/waveform/renderers/waveformbeat.h:
```diff
@@ -0,0 +1,42 @@
+
+using BeatType = mixxx::track::io::Type;
+
+enum Direction {
```
enum class Direction : int

uklotzde commented on src/waveform/renderers/waveformbeat.h:
```diff
@@ -0,0 +1,42 @@
+    void setPosition(int position) {
+        m_position = position;
+    }
+    void setType(BeatType beatType) {
```
Do you really need to modify the type or orientation after construction? Try to make as many members immutable as possible. Dumb setters are considered an anti-pattern.

uklotzde commented on src/waveform/renderers/waveformbeat.h:
```diff
@@ -0,0 +1,42 @@
+    RIGHT
+};
+
+class WaveformBeat {
```
Either declare a virtual destructor or declare this class as `final`. The latter may apply here.

uklotzde commented on src/widget/wbeatmenu.cpp:
```diff
@@ -0,0 +1,11 @@
+    addAction(m_pSetAsDownbeat);
+}
+
+WBeatMenu::~WBeatMenu() {
```
= default (if possible header, otherwise in .cpp file)

uklotzde commented on src/widget/wbeatmenu.cpp:
```diff
@@ -0,0 +1,11 @@
+
+WBeatMenu::WBeatMenu(UserSettingsPointer pConfig, QWidget* parent)
+        : QMenu(parent) {
+    m_pSetAsDownbeat = new QAction(tr("Set as Downbeat"), this);
```
- Initialize in initializer list
- Use parented_ptr/make_parented

uklotzde commented on src/widget/wbeatmenu.h:
```diff
@@ -0,0 +1,24 @@
+  public:
+    WBeatMenu(UserSettingsPointer pConfig, QWidget* parent = nullptr);
+
+    virtual ~WBeatMenu() override;
```
Redundant virtual keyword

uklotzde commented on src/widget/wbeatmenu.h:
```diff
@@ -0,0 +1,24 @@
+    void setTrack(const TrackPointer& track) {
+        m_pTrack = track;
+    }
+  signals:
```
Add separating blank line

uklotzde commented on src/widget/wwaveformviewer.cpp:
```diff
@@ -156,6 +164,12 @@ void WWaveformViewer::mouseMoveEvent(QMouseEvent* event) {
             }
         }
+        BeatPointer pBeat = m_pHoveredBeat = m_waveformWidget->getBeatAtPoint(event->pos());
+        if (pBeat) {
```
This if/else is redundant

uklotzde commented on src/widget/wwaveformviewer.cpp:
```diff
@@ -286,3 +301,8 @@ void WWaveformViewer::unhighlightMark(WaveformMarkPointer pMark) {
 }
+
+void WWaveformViewer::slotDownbeatUpdated() {
+    m_waveformWidget->getTrackInfo()->getBeats()->setDownbeatOffset(m_pHoveredBeat->getIndex() % 4);
```
Replace literal with the suggested kDefaultBeatsPerMeasure

hacksdump committed:
scons: add dependencies

hacksdump committed:
beat: set default instead of curly braces

hacksdump committed:
downbeats: add persistence

Persistence works fine for both beatgrid and beatmap.
But the deserialization result for beatmap's downbeat index is
overriden by the analyzer thread which is currently incapable
of generating downbeat information.

The downbeat restore on mixxx reopen works fine for beatgrid though.

This behavior can be fixed with the #2512 merge or with downbeat
detection.

hacksdump committed:
beat: remove shared pointer

hacksdump committed:
remove redundant virtual declaration

hacksdump committed:
use named constant for common time signature

hacksdump committed:
waveformbeat: use enum class for direction

hacksdump committed:
waveformbeat: make class final

hacksdump committed:
wwaveformviewer: remove redundant assignment

hacksdump committed:
wbeatmenu: default destructor in header

hacksdump committed:
wbeatmenu: use parented_ptr for QAction

hacksdump committed:
depends.py: add missing comma

hacksdump committed:
beatmap: remove stray debug statements

Be-ing commented on src/preferences/dialog/dlgprefwaveform.cpp:
```diff
@@ -29,6 +29,10 @@ DlgPrefWaveform::DlgPrefWaveform(QWidget* pParent, MixxxMainWindow* pMixxx,
 
+    // Populate beatgrid options.
+    beatGridModeComboBox->addItem(tr("Beats"));
+    beatGridModeComboBox->addItem(tr("Beats + Downbeats"));
```
Read the value of the setting from the config file and select the corresponding combobox item here

Be-ing replied:
Oh I see the combobox item is set in slotUpdate. However, the config value is not read on startup and  `WaveformWidgetFactory::instance()->setBeatGridMode` is not called until changed settings are applied.

Holzhaus replied:
I think you can just call slotUpdate in the constructor.

hacksdump replied:
The call already exists. And I checked boot configuration, everything seems to load in place.

`mixxx.cpp` Line 448
```cpp
    WaveformWidgetFactory::instance()->setConfig(pConfig);
```

Be-ing replied:
I tested again, and yes the combobox in DlgPrefWaveform is loaded correctly. However, I cannot get downbeats to show on the waveform until I open the preferences and change the combobox value and that has to be repeated every time I start Mixxx.

Be-ing commented on src/preferences/dialog/dlgprefwaveform.cpp:
```diff
@@ -158,6 +167,9 @@ void DlgPrefWaveform::slotResetToDefaults() {
     defaultZoomComboBox->setCurrentIndex(3 + 1);
 
+    // Default Beat grid mode is "beats + downbeats"
+    beatGridModeComboBox->setCurrentIndex(1);
```
Create an enum class for these index values

Be-ing commented on src/waveform/renderers/waveformbeat.h:
```diff
@@ -0,0 +1,46 @@
+  public:
+    WaveformBeat();
+    void draw(QPainter* painter) const;
+    void setPosition(int position) {
```
What are the units of this position? Audio frames? Pixels?

Be-ing commented on src/waveform/renderers/waveformbeat.h:
```diff
@@ -0,0 +1,46 @@
+    void setLength(int length) {
+        m_length = length;
+    }
+    void setBeatGridMode(int mode) {
```
Use the same enum class here as will be used in DlgPrefWaveform

Be-ing commented on src/waveform/renderers/waveformbeat.cpp:
```diff
@@ -0,0 +1,72 @@
+    }
+}
+
+QPolygonF WaveformBeat::getEquilateralTriangle(float edgeLength,
```
It looks like this doesn't depend on any class members, so it can be moved to the anonymous namespace instead of being a private class member.

Be-ing commented on src/waveform/renderers/waveformrenderbeat.cpp:
```diff
@@ -76,9 +76,11 @@ void WaveformRenderBeat::draw(QPainter* painter, QPaintEvent* /*event*/) {
-        double beatPosition = it->next();
+        auto beat = it->next();
+        double beatPosition = beat.getSamplePosition();
         double xBeatPoint =
                 m_waveformRenderer->transformSamplePositionInRendererWorld(beatPosition);
```
Can we come up with a better description than "x"? What is "x"? A position within the widget? What are the units? Pixels?

Be-ing commented on src/waveform/renderers/waveformrenderbeat.cpp:
```diff
@@ -89,13 +91,25 @@ void WaveformRenderBeat::draw(QPainter* painter, QPaintEvent* /*event*/) {
         } else {
-            m_beats[beatCount++].setLine(0.0f, xBeatPoint, rendererWidth, xBeatPoint);
+            m_beats[beatCount].setLength(rendererWidth);
+            m_beats[beatCount].setOrientation(Qt::Vertical);
```
How about moving this out of the if/else block and always explicitly setting the orientation from the waveform's orientation? That way it does not rely on an implicit default value.

Be-ing commented on src/waveform/renderers/waveformwidgetrenderer.cpp:
```diff
@@ -305,3 +309,25 @@ WaveformMarkPointer WaveformWidgetRenderer::getCueMarkAtPoint(QPoint point) cons
+            }
+        }
+    }
+    // Maybe consider using std::optional return type for this function.
```
Yes I think std::optional would be appropriate here.

Be-ing commented on src/waveform/renderers/waveformwidgetrenderer.cpp:
```diff
@@ -305,3 +309,25 @@ WaveformMarkPointer WaveformWidgetRenderer::getCueMarkAtPoint(QPoint point) cons
+    for (const auto& pBeat : m_beatPositions.keys()) {
+        float position = m_beatPositions[pBeat];
+        if (getOrientation() == Qt::Horizontal) {
+            if (position - 5 < point.x() && point.x() < position + 5) {
```
What is this magic number 5? I presume it's a padding width on either side of the line to make a wider area to right click? Please move it to a named constant in an anonymous namespace.

hacksdump replied:
That certainly should be a constant
But I would like to point out a design inconsistency that has been introduced. If you look at the function just above this, it happens to be `getCueMarkAtPoint` which returns a `WaveformMarkPointer`, which is the graphic object for the mark.
But the return type of current function is not WaveformBeat / WaveformBeatPointer but the Beat object itself.
This means I can't currently use the `contains` method used with WaveformMarkPointer to check if point is in vicinity.
To actually start using WaveformBeat insead of Beat, I would need to store the beat object inside the WaveformBeat class.

A positive side for making this consistent would be:
We can create an interface like `WaveformMarkRightClickable` (for the lack of a good name) which has a method `contains` and `WaveformMark` and `WaveformBeat` both have to implement that interface. I mean pattern comes from Java, I'd like to hear your suggestions on how to do it in C++.

Be-ing replied:
That seems like a good idea. Then we could have less special cases to handle in WWaveformWidgetRenderer.

Be-ing commented on src/waveform/renderers/waveformwidgetrenderer.cpp:
```diff
@@ -305,3 +309,25 @@ WaveformMarkPointer WaveformWidgetRenderer::getCueMarkAtPoint(QPoint point) cons
+    // Maybe consider using std::optional return type for this function.
+    // Optional wrapper should be suitable for searches.
+    Beat beat;
+    // Currently relying on setting enabled status to false.
+    beat.setEnabled(false);
```
I was confused what this "enabled" state is for in the new Beat class. Is this all it is used for? If so, please get rid of it and use std::optional here.

hacksdump replied:
The `enabled` state is not introduced by the new Beat class, it was existing in the Beat protobuf. Considering this check, I will wrap the return type with std::optional.

Be-ing replied:
I have no idea what it means for a beat to be "enabled"... was this previously used in the code anywhere? If not, let's remove it.

Be-ing commented on src/waveform/waveformwidgetfactory.cpp:
```diff
@@ -579,7 +583,17 @@ void WaveformWidgetFactory::setDisplayBeatGridAlpha(int alpha) {
+
+void WaveformWidgetFactory::setBeatGridMode(int mode) {
+    m_beatGridMode = mode;
+    if (m_waveformWidgetHolders.size() == 0) {
+        return;
+    }
```
Is this check necessary? Shouldn't the loop below go through 0 iterations if m_waveformWidgetHolders is empty?

Be-ing commented on src/widget/wwaveformviewer.cpp:
```diff
@@ -91,6 +98,12 @@ void WWaveformViewer::mousePressEvent(QMouseEvent* event) {
+            if (m_hoveredBeat.isEnabled()) {
+                m_pBeatMenu->update();
+                m_pBeatMenu->popup(event->globalPos());
+            }
```
Hmm... how should we handle this situation?
![two-menus](https://user-images.githubusercontent.com/9455094/84335650-f12ebd80-ab5a-11ea-854e-76e0935634b0.png)

Be-ing replied:
I have two ideas how this could be handled:
1. Show WBeatMenu and add QAction "Edit cue" that shows WCueMenuPopup.
2. Add actions to WCueMenuPopup for each of WBeatMenu's actions.

I'm leaning towards 1. What do others think? @Holzhaus, @ronso0?

hacksdump replied:
I kinda had 1 in mind earlier since I was aware of this situation. I was just thinking it wouldn't be appropriate to move the cue menu as an action in BeatMenu since a cue can also be set at a position where there is no beat (if quantization is disabled).

Be-ing replied:
Right, this only only be for the special case where a cue and beat are at the same point (which is a common scenario with quantize). For cues that are not at beats, simply show WCueMenuPopup by itself like is done currently.

Holzhaus replied:
1. requires an additional click, but I think it makes sense here.

Be-ing commented on src/widget/wwaveformviewer.cpp:
```diff
@@ -286,3 +305,8 @@ void WWaveformViewer::unhighlightMark(WaveformMarkPointer pMark) {
     return m_pPlayEnabled->get();
 }
+
+void WWaveformViewer::slotDownbeatUpdated() {
+    m_waveformWidget->getTrackInfo()->getBeats()->setDownbeatStartIndex(
+            m_hoveredBeat.getIndex() % kDefaultBeatsPerMeasure);
+}
```
This feels hacky to do this in WWaveformViewer. I propose moving this into WBeatMenu and setting the active Beat from WWaveformViewer, similar to how we set the active Cue for WCueMenuPopup.

Be-ing commented on src/widget/wbeatmenu.cpp:
```diff
@@ -0,0 +1,16 @@
+
+void WBeatMenu::update() {
+    // TODO(hacksdump): Declare enum type for beat grid mode setting.
+    int beatMode = m_pConfig->getValue(ConfigKey("[Waveform]", "beatGridMode"), 0);
```
I don't think WBeatMenu should be accessing configuration values... how about updating the beatgrid mode for WBeatMenu from WWaveformViewer::setBeatGridMode?

hacksdump committed:
dlgprefwaveform: add BeatGridMode enum

hacksdump committed:
enum BeatGridMode: add comment

hacksdump committed:
enum BeatGridMode: use enum class

hacksdump committed:
waveformwidgetfactory: remove redundant size check

hacksdump committed:
create interface for clickable items on waveform

hacksdump committed:
waveformwidgetrenderer: store list (not map) for beats

hacksdump committed:
beats.proto: remove unused enabled state

hacksdump committed:
beatmenu: update downbeat from this class

hacksdump committed:
beatmenu: hide set downbeat option for downbeat

hacksdump committed:
beatmenu: add configurable options

hacksdump committed:
waveformbeat: move triangle generator to anonymous namespace

hacksdump committed:
waveformrenderbeat: more informative variable name

hacksdump committed:
waveformbeat: descriptive function name

hacksdump committed:
wbeatmenu: handle beat and cue at same position

hacksdump committed:
wwaveformviewer: menu visibility check

Be-ing commented on src/waveform/renderers/waveformbeat.cpp:
```diff
@@ -74,3 +74,19 @@ QPolygonF WaveformBeat::getEquilateralTriangle(float edgeLength,
+
+bool WaveformBeat::contains(QPoint point, Qt::Orientation orientation) const {
+    Q_UNUSED(orientation);
+    return (m_orientation == Qt::Horizontal &&
+                   (m_position - kClickableLinePaddingPixels < point.x() &&
+                           point.x() <
+                                   m_position + kClickableLinePaddingPixels)) ||
+            (m_orientation == Qt::Vertical &&
+                    (m_position - kClickableLinePaddingPixels < point.y() &&
+                            point.y() <
+                                    m_position + kClickableLinePaddingPixels));
```
These heavily nested conditional statements are hard to read. Can you check the orientation in an if/else block?

Holzhaus replied:
And and maybe use a temp var to do the kClickableLinePaddingPixels addition/subtraction only once.

Be-ing commented on src/waveform/renderers/waveformrenderbeat.cpp:
```diff
@@ -81,18 +81,18 @@ void WaveformRenderBeat::draw(QPainter* painter, QPaintEvent* /*event*/) {
         double beatPosition = beat.getSamplePosition();
-        double xBeatPoint =
+        double beatPixelPositionInWidgetSpace =
                 m_waveformRenderer->transformSamplePositionInRendererWorld(beatPosition);
```
Use qRound on this line, then we don't need a separate line just for rounding.

Be-ing commented on src/widget/wbeatmenu.cpp:
```diff
@@ -10,6 +10,11 @@ WBeatMenu::WBeatMenu(UserSettingsPointer pConfig, QWidget* parent)
+    QMenu::popup(pos, at);
+}
+
 void WBeatMenu::update() {
```
This is easy to confuse with QWidget::update, which is used to redraw widgets. Maybe call this updateFlags?

hacksdump replied:
This function does not upload the flags, it updates the menu. So let's call it `updateMenu`

Be-ing commented on src/widget/wbeatmenu.h:
```diff
@@ -6,18 +6,24 @@
 
     ~WBeatMenu() override = default;
 
+    // This function hides (not overrides) non-virtual QMenu::popup().
+    // This has been done on purpose to update menu before showing.
+    void popup(const QPoint& pos, QAction* at = nullptr);
```
I think it would be better to update the QActions when the option flags are changed instead of having this confusing hiding of a standard QMenu method.

Be-ing commented on src/widget/wbeatmenu.cpp:
```diff
@@ -18,6 +23,11 @@ void WBeatMenu::update() {
         addAction(m_pSetAsDownbeat);
     }
+    if (m_eSelectedOptions.testFlag(WBeatMenu::Option::CueMenu)) {
+        m_pCueMenu = make_parented<QAction>(tr("Cue Menu"), this);
```
```suggestion
        m_pCueMenu = make_parented<QAction>(tr("Edit cue"), this);
```

Be-ing commented:
It would be great if you could increase the maximum that the waveforms can be zoomed out to and hide beats that aren't downbeats beyond the current maximum zoom level. I set the current maximum where it is because zooming it out further creates visual chaos with the beats so close together.

Be-ing commented:
Thanks for the small commits. It makes reviewing easy.

hacksdump commented:
> Thanks for the small commits. It makes reviewing easy.

It makes working easy for me as well, keeps ideas organized. Commits should be atomic.

hacksdump commented:
Why is the build failing with only ubuntu CMake?

Be-ing commented:
EngineBufferScaleLinearTest.UnityRateIsSamplePerfect failed. That was probably a spurious test failure... https://mixxx.zulipchat.com/#narrow/stream/109171-development/topic/Randomly.20failing.20tests

hacksdump committed:
waveformbeat: simplify conditional

hacksdump committed:
waveformrenderbeat: use int type for rounded value

hacksdump committed:
wbeatmenu: remove popup method hiding

hacksdump committed:
wbeatmenu: rename update to updateMenu

hacksdump committed:
wbeatmenu: rename action

Holzhaus commented:
I restarted the job and now I passes.

hacksdump committed:
waveformbeat: increase max zoom | hide beats on zoom out

hacksdump commented:
> It would be great if you could increase the maximum that the waveforms can be zoomed out to and hide beats that aren't downbeats beyond the current maximum zoom level. I set the current maximum where it is because zooming it out further creates visual chaos with the beats so close together.

Check it out :)

An issue is, now that the audio samples are much closer, the scratch sound when dragging with GUI is louder.

hacksdump commented:
Closing in favour of #2961.
